### DISCUSSION TITLE
Deactivate matrix blocks per site

### DISCUSSION THREAD

**Discussion by mihob** [craftcms/cms#10830] [Q&A] [answered]
We have a multilingual site and use a matrix field to build the content of each page with blocks. The blocks are stored in each language.

From time to time, however, it would be desirable to disable blocks in individual languages. But this seems not to be possible, blocks are then always deactivated for all languages.

Can this be done somehow? Or is it basically not possible?

**Comment by brandonkelly** [comment, marked as answer]
That’s currently not possible, however you could pull it off by adding a custom “Enabled for site” Lightswitch field to each block type (defaults to enabled), and set it to be translatable for each site.

Then when querying Matrix blocks, only fetch the ones where that field is enabled:

```twig
{% set blocks = entry.myMatrixField
  .myEnabledForSiteHandle(true)
  .all() %}
```

**Reply by mihob** [reply]
Thanks!

**Comment by mihob** [comment]
@brandonkelly Sorry for my late reply, but I'm a little lost here right now.

We get our content via GraphQL and it doesn't seem possible to filter the MatrixBlocks accordingly.

Our query looks something like this:

```
query Page ($siteId: [QueryArgument]!, $uri: [String]!){
  entry(siteId: $siteId , uri: $uri){
    title
    ... on pages_contentpage_Entry{
      elements(enabledForSite: true) {
        title
      }      
    }
  }
}
```
which results in the following error:

```
{
  "errors": [
    {
      "message": "Unknown argument \"enabledForSite\" on field \"elements\" of type \"pages_contentpage_Entry\".",
      "extensions": {
        "category": "graphql"
      },
      "locations": [
        {
          "line": 5,
          "column": 16
        }
      ]
    }
  ]
}
```
It works without the inline fragment `... on pages_contentpage_Entry`, but i dont really know why it causes an error.

**Reply by brandonkelly** [reply]
Did you name the Matrix block sub-field `enabledForSite`? You may need to come up with a slightly different name (e.g. `blockEnabledForSite`) so it doesn’t possibly conflict with the existing `enabledForSite` param.

**Reply by mihob** [reply]
No, the field is actually named `availableForLanguage`, so there shouldn't be a conflict as far as i know.

**Reply by brandonkelly** [reply]
OK, but your example GraphQL query only has `enabledForSite: true`, not `availableForLanguage: true`.

**Reply by mihob** [reply]
This was just a simplified example and I based it on the field name you gave me.

In reality we have a few more entry types and not everyone has the matrix field `elements`. But every block of `elements `has the field `availableForLanguage`

**Reply by brandonkelly** [reply]
Do you have multiple Matrix block types? In that case you would need to give each `availableForLanguage` field a unique handle. And make sure they are enabled by default. Then list each one in the arguments – `typeA_enabledForLanguge: true, typeB_enabledForLanguage: true`

**Reply by mihob** [reply]
Yes, we have about 20 block types. This would make the queries very unwieldy.

But why do i need to do this? Is there no way to filter blocks by field values in graph ql just like you twig example?

**Reply by brandonkelly** [reply]
That’s exactly what I’m suggesting you do.

**Reply by mihob** [reply]
Hmm, then I don't understand you at all.

Why exactly do I need a different handle for each of my 20 blocks for the same field?

Your twig example suggests that the handle can always be the same for each block.

**Reply by brandonkelly** [reply]
I was asking if you have multiple block _types_ (defined in the Matrix field settings); not whether there were multiple blocks within the entry. I brought that up because if you do, it’s best to ensure each block type has unique sub-field handles, to avoid name conflicts when trying to filter the blocks from a query.

**Reply by mihob** [reply]
Well, there are about 20 different block **types** for the matrix field. So the resulting query would be huge.

Ist there really no simple way to filter matrix field blocks in GraphQl by their field values? Or do i get the field naming wrong?
Maybe Another example helps, some of my block types have a field with the handle `layout`. Now i want to find all blocks where `layout` has the value "1":

```
{
    entries(section: "pages", type: "contentpage" ) {
      sectionHandle
      typeHandle
      elements(layout: "1"){
        typeHandle
      	layout
      }
  }
}
```
The query would not find any blocks even if there were some with a layout value of "1"

And i still don't understand, why using an inline fragment in that context causes an error.

**Reply by brandonkelly** [reply]
> Ist there really no simple way to filter matrix field blocks in GraphQl by their field values?

To be clear, yes, in general it is possible, but there are limitations. It doesn’t work reliably when two or more sub-fields have the same field handle. And in general you can’t filter by fields across multiple block types simultaneously, because `or` conditions aren’t possible over GraphQL.

The only exception to that is Lightswitch fields. If you have, say, two block types, each with a nested Lightswitch field (with unique handles), and _their default values are both set to enabled_, then you could do:

```graphql
myMatrixFieldHandle(lightswitchFieldA: true, lightswitchFieldB: true)
```

That’s only possible for Lightswitch fields because they will treat all `null` field values as whatever their default value is set to. So for example, blocks of type `A` won’t be excluded even though they don’t have `lightswitchFieldB` assigned to them, so long as the field’s default value is enabled. And blocks of type `B` won’t be excluded even though they don’t have `lightswitchFieldA` assigned to them.

**Reply by brandonkelly** [reply]
> Well, there are about 20 different block types for the matrix field. So the resulting query would be huge.

Yes, it will be an ugly query. I didn’t realize you had 20 block types when I made the initial suggestion.

Maybe a better way to go would be to just include these `availabelForLanguage` fields in the output, and do the filtering from JavaScript instead.

**Reply by mihob** [reply]
All right, many thanks for your help!

We are now filtering the blocks in the frontend as you suggested.

Out of curiosity, do you plan to eventually make matrix blocks deactivatable by language/site, analog to entries?

**Reply by brandonkelly** [reply]
We probably will at some point, but no specific plan of attack yet.
